lib/Source/AutoHog/AutoHogImporter.cpp

```diff
@@ -0,0 +1,516 @@
+    auto &connections = cell["connections"];
+    assert(connections.IsObject() && "Expected 'connections' to be an object");
+    for (rapidjson::Value::ConstMemberIterator itr = connections.MemberBegin();
+         itr != connections.MemberEnd(); ++itr) {
+      const char *localPortName = itr->name.GetString();
+      const rapidjson::Value &portData = itr->value;
+      assert(portData.IsObject() &&
+             "Expected 'connections[...]' to be an object");
+      bool isInput = strcmp(cell["port_directions"][localPortName].GetString(),
+                            "input") == 0;
+      if (portData.HasMember("cell") && isInput) {
+        const char *connectedCell = portData["cell"].GetString();
+        // Since the input is not toposorted, we may need to add the incident
+        // vertex.
+        if (!cellGraph.contains(std::string(connectedCell))) {
+          cellGraph.addVertex(std::string(connectedCell));
+        }
+        LLVM_DEBUG(llvm::dbgs() << "Adding edge from " << cellName << " to "
+                                << connectedCell << "\n");
+        cellGraph.addEdge(std::string(cellName), std::string(connectedCell));
+      }
+    }
+  }
+  return cellGraph;
+}
+
+/// Parse a list of integers into a single LUT integer, with the first entry
+/// in the list being the LUT value for the input 0.
+///
+/// The output integer X is interpreted as a LUT via f(input) = (X >> input) & 1
+///
+/// E.g., [1, 0, 1, 0] -> 5
+///
+int32_t parseLut(const rapidjson::GenericArray<true, rapidjson::Value> &lut) {
+  int32_t result = 0;
+  for (int32_t i = 0; i < lut.Size(); i++) {
+    int32_t bit = lut[i].GetInt();
+    result |= bit << i;
+  }
+  return result;
+}
+
+// Parse the coefficients field of a cell object in the JSON document.
+// Populates the result vector with the coefficients in the order of the
+// operands.
+//
+// Args:
+// - coefficients: a map from port name to coefficient value.
+// - inputPortNameToOperandIndex: a map from port name to the
+//   index of the operand in the operands vector.
+// - result: the pre-resized vector to populate with the coefficients.
+void parseLinCombCoefficients(
+    const rapidjson::GenericObject<true, rapidjson::Value> &coefficients,
+    const StringMap<int> &inputPortNameToOperandIndex,
+    SmallVector<int> &result) {
+  for (rapidjson::Value::ConstMemberIterator itr = coefficients.MemberBegin();
+       itr != coefficients.MemberEnd(); ++itr) {
+    const char *portName = itr->name.GetString();
+    const int coeffValue = itr->value.GetInt();
+    assert(inputPortNameToOperandIndex.contains(portName) &&
+           "Coefficient port name not found in inputPortNameToOperandIndex");
+    result[inputPortNameToOperandIndex.at(portName)] = coeffValue;
+  }
+}
+
+OwningOpRef<Operation *> translateFromAutoHog(llvm::StringRef inputString,
+                                              MLIRContext *context) {
+  // FIXME: shouldn't have to manually load these dialects!
```

Comment:
Ah, this might be a bug in `TranslateToMLIRRegistration`?

Reply:
lol this will be the third time I fix dialect loading errors in mlir-translate

Reply:
The dialects are loaded lazily, when something from that dialect is parsed. Nothing in the input is MLIR in this case.